libraries/rush-lib/src/logic/operations/ShardedPhaseOperationPlugin.ts

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import type { IPhase } from '../../api/CommandLineConfiguration';
+import { EnvironmentVariableNames } from '../../api/EnvironmentConfiguration';
+import type {
+  ICreateOperationsContext,
+  IPhasedCommandPlugin,
+  PhasedCommandHooks
+} from '../../pluginFramework/PhasedCommandHooks';
+import { NullOperationRunner } from './NullOperationRunner';
+import { Operation } from './Operation';
+import { OperationStatus } from './OperationStatus';
+import { ShellOperationRunner } from './ShellOperationRunner';
+import {
+  formatCommand,
+  getCustomParameterValuesByPhase,
+  getDisplayName,
+  getScriptToRun
+} from './ShellOperationRunnerPlugin';
+
+export const PLUGIN_NAME: 'ShardedPhasedOperationPlugin' = 'ShardedPhasedOperationPlugin';
+
+// eslint-disable-next-line @typescript-eslint/typedef
+const TemplateStrings = {
+  SHARD_INDEX: '{shardIndex}',
+  SHARD_COUNT: '{shardCount}'
+} as const;
+
+// eslint-disable-next-line @typescript-eslint/typedef
+const TemplateStringRegex = {
+  SHARD_INDEX: new RegExp(TemplateStrings.SHARD_INDEX, 'g'),
+  SHARD_COUNT: new RegExp(TemplateStrings.SHARD_COUNT, 'g')
+} as const;
+
+/**
+ * Phased command that shards a phase into multiple operations.
+ */
+export class ShardedPhasedOperationPlugin implements IPhasedCommandPlugin {
+  public apply(hooks: PhasedCommandHooks): void {
+    hooks.createOperations.tap(PLUGIN_NAME, spliceShards);
+  }
+}
+
+function spliceShards(existingOperations: Set<Operation>, context: ICreateOperationsContext): Set<Operation> {
+  const { rushConfiguration } = context;
+
+  const getCustomParameterValuesForPhase: (phase: IPhase) => ReadonlyArray<string> =
+    getCustomParameterValuesByPhase();
+
+  for (const operation of existingOperations) {
+    const { associatedPhase: phase, associatedProject: project, settings: operationSettings } = operation;
+    if (phase && project && operationSettings?.sharding && !operation.runner) {
```

Comment:
So if an operation runner is already defined, the sharding operation settings are ignored?

Reply:
Yeah, I was following what [`ShellOperationRunnerPlugin`](https://github.com/aramissennyeydd/rushstack/blob/42f63c2d88fea68022699345055e59b1036c9789/libraries/rush-lib/src/logic/operations/ShellOperationRunnerPlugin.ts#L40) and [`IPCOperationRunnerPlugin`](https://github.com/aramissennyeydd/rushstack/blob/42f63c2d88fea68022699345055e59b1036c9789/libraries/rush-lib/src/logic/operations/IPCOperationRunnerPlugin.ts#L50) do.